cpp/src/join/distinct_hash_join.cu

```diff
@@ -328,14 +343,50 @@ distinct_hash_join<HasNested>::inner_join(rmm::cuda_stream_view stream,
     probe_table_num_rows,
     this->_hash_table.ref(cuco::find),
     counter.data(),
-    left_indices->data(),
-    right_indices->data());
+    build_indices->data(),
+    probe_indices->data());
 
   auto const actual_size = counter.value(stream);
-  left_indices->resize(actual_size, stream);
-  right_indices->resize(actual_size, stream);
+  build_indices->resize(actual_size, stream);
+  probe_indices->resize(actual_size, stream);
+
+  return {std::move(build_indices), std::move(probe_indices)};
+}
+
+template <cudf::has_nested HasNested>
+std::unique_ptr<rmm::device_uvector<size_type>> distinct_hash_join<HasNested>::left_join(
+  rmm::cuda_stream_view stream, rmm::mr::device_memory_resource* mr) const
+{
+  cudf::thread_range range{"distinct_hash_join::left_join"};
+
+  size_type const probe_table_num_rows{this->_probe.num_rows()};
+
+  // If output size is zero, return empty
+  if (probe_table_num_rows == 0) {
+    return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+  }
+
+  auto build_indices =
+    std::make_unique<rmm::device_uvector<size_type>>(probe_table_num_rows, stream, mr);
+
+  // If build table is empty, return probe table
+  if (this->_build.num_rows() == 0) {
+    thrust::fill(
+      rmm::exec_policy_nosync(stream), build_indices->begin(), build_indices->end(), JoinNoneValue);
+  } else {
+    auto const probe_row_hasher =
+      cudf::experimental::row::hash::row_hasher{this->_preprocessed_probe};
+    auto const d_probe_hasher = probe_row_hasher.device_hasher(nullate::DYNAMIC{this->_has_nulls});
+    auto const iter           = cudf::detail::make_counting_transform_iterator(
+      0, build_keys_fn<decltype(d_probe_hasher), rhs_index_type>{d_probe_hasher});
+
+    auto const output_begin =
+      thrust::make_transform_output_iterator(build_indices->begin(), output_fn{});
+    // TODO conditional find for nulls
```

Comment:
Should this TODO be done first?

Reply:
This depends on `static_set.find_if_async()` which hasn't been added in cuco yet.

Reply:
I've updated the comment to provide more details about the TODO task [00afba3](https://github.com/rapidsai/cudf/pull/15149/commits/00afba3408f5d335bd11aecb760a1e690a1f9c63)